Print using callbacks (#621)
PennyLaneAI/catalyst: doc/changelog.md

  [(#650)](https://github.com/PennyLaneAI/catalyst/pull/650)
  [(#649)](https://github.com/PennyLaneAI/catalyst/pull/649)
  [(#661)](https://github.com/PennyLaneAI/catalyst/pull/661)
  [(#621)](https://github.com/PennyLaneAI/catalyst/pull/621)

  Catalyst now supports callbacks with parameters and return values.
  The following is now possible:

  ```py
  @debug.callback
  def foo(val):

>     return val

Author: ```suggestion
    print(val)
```

>     return val

Reviewer: This doesnt showcase the "debug" aspect very well, maybe an f-string print would be nice?

Author: Ah yes I did `git push -f` after the update and didn't update what I had done in the online code editor.

  @pure_callback
  def bar(val) -> int:
    return val + 1

  @qjit
  def circuit(param):
    x = bar(param)
    foo(x)

  ```

  ```pycon
  >>> print(circuit(123))

>   124

Author: ```suggestion
myval: 124
```

  ```

  This includes support for the specialized `pure_callback` and `debug.callback` where
  `pure_callback` is expected to return a value and be side effect free,
  while `debug.callback` is expected to produce a side effect and have no return values.

  Syntactic sugar on top of `debug.callback` includes `debug.print` which allows the user

>   to use f-stringlike to format the string before printing.

Reviewer: I would compare it `str.format()` rather than fstrings because its behaviour is closer to that right? 
`print("{c} {b} {a}", a=a, b=b, c=c)` ~ `print("{c} {b} {a}".format(a=a, b=b, c=c))`

Reviewer: Might even make sense to link to the doc page: https://docs.python.org/3/library/string.html#formatstrings

Author: ```suggestion
  to use formatstrings to format the string before printing. See https://docs.python.org/3/library/string.html#formatstrings
```

  ```py
  @qjit
  def cir(a, b, c):
      debug.print("{c} {b} {a}", a=a, b=b, c=c)
  ```

  ```pycon
  >>> cir(1, 2, 3)
  3 2 1

>   ```

Reviewer 2: This will be a massive QOL improvement 😎

  123
  ```

  There is also a callback to improve the interface for printing.
  It is equivalent to JAX's `debug_print` function. E.g.,

  ```py
  @qml.qjit
  def func(*args, **kwargs):
     debug_print("{arg2}, {arg1}, arg2=args[1], args1=[0])
  ```

  ```pycon

>   >>> print(func(1, 2))
>   2 1

Reviewer 2: ```suggestion
  >>> func(1, 2)
  2 1
```

since 'console syntax' will always automatically print output :)

(technically, it will print the `repr()` rather than the `str()`, but same effect).

Author: From the testing, it looks like print uses string rather than repr : https://stackoverflow.com/a/36156152

Author: https://github.com/PennyLaneAI/catalyst/pull/621/commits/09f95c5aa3ce6575dce1b3374b2e11a5e043c561

<h3>Breaking changes</h3>

> * The print function has changed. As a result, the memref keyword argument has been removed.

Author: Changed here. Thanks! https://github.com/PennyLaneAI/catalyst/pull/621/commits/06f375e6667f47519fee81f196323a920c5c0b20
